lib/src/main/kotlin/trees/abstractTree.kt

```diff
@@ -0,0 +1,157 @@
+package trees
+
+import nodes.abstractNode
+
+abstract class abstractTree<K: Comparable<K>, V, someNode: abstractNode<K, V, someNode>> {
+    protected var root: someNode? = null
+
+    protected abstract fun createNewNode(key: K, value: V): someNode
+
+    open fun insert(key: K, value: V) {
+        insertNode(key, value)
+    }
+
+    protected fun insertNode(key: K, value: V): someNode? {
+        val nodeToInsert = createNewNode(key, value)
+        var curNode = root
+
+        if (curNode == null) {
+            root = nodeToInsert
+            return root
+        }
+
+        while ((curNode != null) && (curNode != nodeToInsert)) {
+            when {
+                (curNode.key == nodeToInsert.key) -> {
+                    curNode.value = nodeToInsert.value
+                    break
+                }
+
+                ((curNode.leftChild == null) && (nodeToInsert.key < curNode.key)) -> {
+                    curNode.leftChild = nodeToInsert
+                    curNode = curNode.leftChild
+                }
+
+                ((curNode.rightChild == null) && (nodeToInsert.key > curNode.key)) -> {
+                    curNode.rightChild = nodeToInsert
+                    curNode = curNode.rightChild
+                }
+
+                nodeToInsert.key < curNode.key -> {
+                    curNode = curNode.leftChild
+                }
+
+                nodeToInsert.key > curNode.key -> {
+                    curNode = curNode.rightChild
+                }
+            }
+        }
+        return curNode
+
+    }
+
+    open fun delete(key: K) {
+        deleteNode(key)
+    }
+
+    protected fun deleteNode(key: K): someNode? {
+        val nodeToDelete = findNodeByKey(key)
+        if ((nodeToDelete == null) || (root == null)) return null
+        val parentNode = findParent(nodeToDelete)
+        if ((nodeToDelete != root) && (parentNode == null)) {
+            throw IllegalArgumentException("Non-root should have parent")
+        }
+        when {
+            /* no children case */
+            (nodeToDelete.leftChild == null && nodeToDelete.rightChild == null) -> {
+                changeChild(nodeToDelete, parentNode, null)
+                return null
+            }
+
+            /* 1 child case */
+            (nodeToDelete.leftChild == null || nodeToDelete.rightChild == null) -> {
+                if (nodeToDelete.leftChild == null) {
+                    changeChild(nodeToDelete, parentNode, nodeToDelete.rightChild)
+                    return nodeToDelete.rightChild
+                }
+                else {
```

Comment:
Тут что-то немного сломалось?
```suggestion
                } else {
```